File: 012_Capstone/consumer.py
from pyspark.streaming.kafka import KafkaUtils
from pyspark.streaming import StreamingContext
from pyspark import SparkContext
from pyspark.sql import SparkSession
from collections import namedtuple 
import pyspark.sql.types as st 
import json
from json import loads
import math

import sys
reload(sys)
sys.setdefaultencoding('utf-8')


sc = SparkContext(appName="capstonePlayStore")                                                                                     
ssc = StreamingContext(sc, 10)   
ks = KafkaUtils.createDirectStream(ssc, ['CapstoneSamirTest'], {'metadata.broker.list': 'localhost:9092'})  

mymongo = SparkSession.builder.appName("capstonePlayStore")\
    .config("spark.mongodb.input.uri", "mongodb://127.0.0.1/test.playstore") \
    .config("spark.mongodb.output.uri", "mongodb://127.0.0.1/test.playstore") \
	.getOrCreate()

mymongo.sparkContext.setLogLevel('WARN') 
  

result1 = ks.map(lambda x: json.loads(x[1]))

fields = ("developerId", "title", "url" , "priceText", "score", "scoreText" )
apps = namedtuple("applications", fields)


def handle_rdd(rdd):                                                                                                    
    if not rdd.isEmpty():                                                                                               
        global mymongo                                                                                 
        df = mymongo.createDataFrame(rdd)  
        df.write.format("mongo").mode("append").save() 
        dfOrderdBy = df.orderBy('score', ascending=False)                               
        dfOrderdBy.show()
    	df.printSchema() 
                                                                                                   
         
data = result1.flatMap(lambda x: json.loads(x))\
    .map(lambda x: apps(x['developerId'], x['title'], x['url'], x['priceText'], x['score'], x['scoreText'] ))\
    .filter(lambda price: price[3] == 'FREE')\
    .filter(lambda numb: numb[4] > 4 )\
    .foreachRDD(lambda x: handle_rdd(x))

print("Well Done Samir...")

ssc.start()                                                                                                             
ssc.awaitTermination()

File: 012_Capstone/consumer.scala
import org.apache.spark.streaming.StreamingContext._
import org.apache.spark._
import org.apache.spark.streaming._
import org.apache.spark.sql._
import org.apache.spark.sql.types._
import org.apache.spark.streaming.kafka._
import java.util.Properties
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkConf
import scala.collection.mutable.ListBuffer

class consumer
{	
	def main(args: Array[String]): Unit =
	{
		val linelist: ListBuffer[String] = new ListBuffer[String]();
	
        val ss = SparkSession.builder.appName("capstonePlayStore")\
                .config("spark.mongodb.input.uri", "mongodb://127.0.0.1/test.playstore") \
                .config("spark.mongodb.output.uri", "mongodb://127.0.0.1/test.playstore") \
                .getOrCreate()


		import ss.implicits._
		val sc = ss.sparkContext
 		val ssc = new StreamingContext(sc, Seconds(10))

		val sqlContext = new org.apache.spark.sql.SQLContext(sc)

		val dstream = KafkaUtils.createStream(ssc, "localhost:9092", "playstore" , Map("data"-> 1))

		
		val mappedlines = dstream.foreachRDD{ rdd =>
      		rdd.foreach{record =>
			     var mylist = List(record._2)
			     print(mylist)
			     var myrdd = sc.parallelize(mylist)		     
			     var myrdd2 = myrdd.map{x => Row( x['developerId'], x['title'], x['url'], x['priceText'], x['score'], x['scoreText'] )}
			     var df = sqlContext.createDataFrame( myrdd2 )
			     
			     if (df != null) {
                        df.write.format("mongo").mode("append").save() 
                    }
			   }
		}
		ssc.start
   		ssc.awaitTermination
	}
}